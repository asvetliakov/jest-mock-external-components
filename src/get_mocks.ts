import { parse } from "@babel/parser";
import traverse, { TraverseOptions, Binding, Scope } from "@babel/traverse";
import * as t from "@babel/types";

export interface Mock {
    identifier: string;
    path: string;
    type: "namespace" | "name" | "default";
}

const createNewIdentifierOrMemberExp = (node: t.Identifier | t.MemberExpression | t.JSXIdentifier | t.JSXMemberExpression): t.Identifier | t.MemberExpression => {
    if (t.isIdentifier(node) || t.isJSXIdentifier(node)) {
        return t.identifier(node.name);
    }
    return t.memberExpression(
        createNewIdentifierOrMemberExp(node.object as t.Identifier | t.MemberExpression),
        t.identifier((node.property as t.Identifier).name),
    );
}
const getMostLeftJSXIdentifier = (exp: t.JSXMemberExpression): t.JSXIdentifier => {
    if (t.isJSXIdentifier(exp.object)) {
        return exp.object;
    }
    return getMostLeftJSXIdentifier(exp.object);
}
const getMostLeftIdentifier = (exp: t.Node): t.Identifier | undefined => {
    if (!t.isMemberExpression(exp)) {
        return;
    }
    if (t.isIdentifier(exp.object)) {
        return exp.object;
    }
    return getMostLeftIdentifier(exp.object);
}
const getFullNameOfIdentifier = (exp: t.JSXMemberExpression | t.JSXIdentifier | t.Identifier | t.MemberExpression): string => {
    if (t.isIdentifier(exp) || t.isJSXIdentifier(exp)) {
        return exp.name;
    }
    return getFullNameOfIdentifier(exp.object as t.MemberExpression | t.Identifier) + `.${(exp.property as t.Identifier).name}`;
}

const processHocLikeCallExpression = (exp: t.CallExpression, scope: Scope): [t.Identifier | t.MemberExpression | undefined, Binding | undefined] => {
    if (!exp.arguments.length) {
        return [undefined, undefined];
    }
    // get first identifier or member exp, for HOC this is most common case, don't want to bother with anothers
    const firstArg: t.MemberExpression | t.Identifier | undefined = exp.arguments.find(a => t.isIdentifier(a) || t.isMemberExpression(a)) as any;
    if (!firstArg) {
        return [undefined, undefined];
    }
    const identifier = t.isIdentifier(firstArg) ? firstArg : getMostLeftIdentifier(firstArg);
    if (!identifier) {
        return [undefined, undefined];
    }
    const binding = scope.getBinding(identifier.name);
    if (!binding) {
        return [undefined, undefined];
    }
    if (binding.kind === "module") {
        // direct binding to module, return used binding
        return [firstArg, binding];
    } else {
        // import B;
        // const A = HOC(B);
        // const C = HOC(A);
        // <C />
        // need to process again
        if (!binding.path.isVariableDeclarator() || !(t.isTaggedTemplateExpression(binding.path.node.init!) || t.isCallExpression(binding.path.node.init!))) {
            return [undefined, undefined];
        }
        const init = binding.path.node.init;
        if (!init) {
            return [undefined, undefined];
        }
        if (t.isTaggedTemplateExpression(init) && !t.isCallExpression(init.tag)) {
            return [undefined, undefined];
        }
        let callExp = t.isCallExpression(init) ? init : (init as t.TaggedTemplateExpression).tag as t.CallExpression;
        // const A = (0, _someHOC)(_ImportedComponent)(" styles ") - often as result of styled components preprocessors
        if (t.isCallExpression(callExp.callee)) {
            callExp = callExp.callee;
        }
        return processHocLikeCallExpression(callExp, binding.path.scope);
    }
}

interface State {
    mocks: Mock[];
    alwaysMockIdentifiers: string[];
    mockedIdentifiers: string[];
}


const visitor: TraverseOptions = {
    CallExpression(path, state: State) {
        const callIdentifier = t.isIdentifier(path.node.callee)
            ? path.node.callee
            : t.isMemberExpression(path.node.callee) && t.isIdentifier(path.node.callee.property)
                ? path.node.callee.property
                : undefined;
        if (!callIdentifier || !state.alwaysMockIdentifiers.includes(callIdentifier.name)) {
            return;
        }
        let identifierNode: t.Identifier | t.MemberExpression | undefined;
        let identifierBinding: Binding | undefined;
        if (t.isCallExpression(path.node.callee)) {
            [identifierNode, identifierBinding] = processHocLikeCallExpression(path.node.callee, path.scope);
        }
        if (!identifierNode) {
            [identifierNode, identifierBinding] = processHocLikeCallExpression(path.node, path.scope);
        }
        if (!identifierNode || !identifierBinding || identifierBinding.kind !== "module") {
            return;
        }
        const identifierFullName = getFullNameOfIdentifier(identifierNode);
        if (state.mockedIdentifiers.includes(identifierFullName)) {
            return;
        }
        // if (!state.mockExpression) {
        //     state.mockExpression = createEmptyMockExpression(identifierBinding.scope);
        // }
        // pushMockObjExpression(this.mockExpression, identifierFullName, identifierBinding);
        state.mocks.push({
            identifier: identifierFullName,
            path: (identifierBinding.path.parent as t.ImportDeclaration).source.value,
            type: identifierBinding.path.isImportDefaultSpecifier() ? "default" : identifierBinding.path.isImportNamespaceSpecifier() ? "namespace" : "name",
        });
        state.mockedIdentifiers.push(identifierFullName);
    },
    JSXOpeningElement(path, state: State) {
        let identifierNode: t.MemberExpression | t.Identifier | t.JSXMemberExpression | t.JSXIdentifier | undefined;
        let identifierBinding: Binding | undefined;
        // <A.B.C /> this can be mapped only to imported module
        if (t.isJSXMemberExpression(path.node.name)) {
            const mostLeftIdentifier = getMostLeftJSXIdentifier(path.node.name);
            const binding = path.scope.getBinding(mostLeftIdentifier.name);
            if (!binding || binding.kind !== "module") {
                return;
            }
            identifierNode = path.node.name;
            identifierBinding = binding;
        } else if (t.isJSXIdentifier(path.node.name)) {
            const identifier = path.node.name.name;
            const binding = path.scope.getBinding(identifier);
            if (!binding) {
                return;
            }
            if (binding.kind === "module") {
                // direct use if imported component
                identifierNode = path.node.name;
                identifierBinding = binding;
            } else {
                // may be hoc
                // 1) const Styled = styled(A)``
                // 2) const A = HOC(B);
                // 3) const C = A.B.C; // member access shorthand
                if (!binding.path.isVariableDeclarator()) {
                    return;
                }
                const init = binding.path.node.init;
                if (init && (t.isIdentifier(init) || t.isMemberExpression(init))) {
                    identifierNode = init;
                    const leftIdentifier = getMostLeftIdentifier(identifierNode);
                    if (leftIdentifier) {
                        const newBinding = binding.path.scope.getBinding(leftIdentifier.name);
                        if (newBinding && newBinding.kind === "module") {
                            identifierBinding = newBinding;
                        }
                    }
                } else if (init && (t.isTaggedTemplateExpression(init) || t.isCallExpression(init))) {
                    if (t.isTaggedTemplateExpression(init) && !t.isCallExpression(init.tag)) {
                        return;
                    }
                    let callExp = t.isTaggedTemplateExpression(init) ? init.tag as t.CallExpression : init;
                    if (t.isCallExpression(callExp.callee)) {
                        // const A = (0, _someHOC)(_ImportedComponent)(" styles ") - often as result of styled components preprocessors
                        // note: need to first process inner call expression, otherwise it may not work - see "Does not import inside in taggle template" test
                        [identifierNode, identifierBinding] = processHocLikeCallExpression(callExp.callee, binding.path.scope);
                    }
                    // otherwise process normally
                    if (!identifierNode && !t.isCallExpression(callExp.callee)) {
                        [identifierNode, identifierBinding] = processHocLikeCallExpression(callExp, binding.path.scope);
                    }
                }
            }
        }
        if (!identifierNode || !identifierBinding || identifierBinding.kind !== "module") {
            return;
        }
        const identifierFullName = getFullNameOfIdentifier(identifierNode);
        if (state.mockedIdentifiers.includes(identifierFullName)) {
            return;
        }
        // if (!this.mockExpression) {
        //     this.mockExpression = createEmptyMockExpression(identifierBinding.scope);
        // }
        // pushMockObjExpression(this.mockExpression, identifierFullName, identifierBinding);
        // this.mockedIdentifiers.push(identifierFullName);
        state.mocks.push({
            identifier: identifierFullName,
            path: (identifierBinding.path.parent as t.ImportDeclaration).source.value,
            type: identifierBinding.path.isImportDefaultSpecifier() ? "default" : identifierBinding.path.isImportNamespaceSpecifier() ? "namespace" : "name",
        });
        state.mockedIdentifiers.push(identifierFullName);
    },
}

export function getMocks(code: string, type: "flow" | "typescript", alwaysMock: string[] = ["styled", "withComponent"]): Mock[] {
    try {
        const state: State = {
            mocks: [],
            mockedIdentifiers: [],
            alwaysMockIdentifiers: alwaysMock,
        };

        const ast = parse(code, {
            sourceType: "module",
            plugins: [
                "jsx",
                type,
                "asyncGenerators",
                "classProperties",
                "classPrivateProperties",
                "classPrivateMethods",
                ["decorators", { decoratorsBeforeExport: true }] as any,
                "doExpressions",
                "dynamicImport",
                "functionBind",
                "functionSent",
                "objectRestSpread",
                "bigInt",
                "exportDefaultFrom",
                "exportNamespaceFrom",
                "importMeta",
                "optionalCatchBinding",
                "optionalChaining",
                "nullishCoalescingOperator"
            ],
        });
        if (!ast) {
            return [];
        }
        traverse(ast, visitor, undefined, state);
        return state.mocks;
    } catch {
        return [];
    }
}