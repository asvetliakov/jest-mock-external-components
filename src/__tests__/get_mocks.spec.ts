import { getMocks } from "../get_mocks";

it("Works for JSX expressions", () => {
    const source = `
        import A from "a";
        import { B } from "../b";
        import * as C from "c/test";

        const Element = () => <div />;

        <A />;
        <A.A1 />;
        <B><div /></B>;
        <B.B1 />;
        <C.C1 />;
        <C.C1.C2 />;
        <C.C2.C3.C4 />;

        class K extends React.Component {
            render() {
                <Element />;
                return (<A><B.B1><C.C1 /></B.B1></A>);
            }
        }

        <K><A /></K>;
        `;

    expect(getMocks(source, "flow")).toMatchSnapshot();
});

it("Works for HOC like components", () => {
    const source = `
        import A from "a";
        import styled from "styled";

        const StyledA = styled(A)\`\`;
        const StyledB = styled(A.B);
        const StyledC = styled(A.C, {});
        const StyledD = styled(A.D, {})\`font-size: 8em;\`;

        <A />;

        class K extends React.Component {
            render() {
                return <div><StyledA /><StyledB /><StyledC /><StyledD /></div>
            }
        }

        <K><StyledA /><StyledB /><StyledC /><StyledD /></K>
        `;
    expect(getMocks(source, "flow")).toMatchSnapshot();
});

it("Does not append imported if not used in HOC as argument", () => {
    const source = `
        import A from "a";
        import styled from "styled";
        import C from "../c";

        const StyledA = styled(A)\`\`;
        const Comp = styled("div")\`\`;
        const StyledB = A;

        const K = C.A.B;


        <StyledA />;
        <Comp />;
        <StyledB />;
        <K />;
        `;
    expect(getMocks(source, "flow")).toMatchSnapshot();
});

it("Does not import inside in tagged template", () => {
    const source = `
        import A from "a";
        import styled from "styled";
        import theme from "../theme";
        import { someFunc, someVal } from "./some";

        const StyledA = styled(A)\`
            color: \${theme.color.black};
        \`;

        const StyledB = styled("div") \`
            height: \${theme.elements.someHeight};
            width: \${theme.elements.someWidth};
        \`;

        const StyledC = styled("div") \`\${theme.someFunc()}\`;
        const StyledD = styled("div") \`\${someFunc()}\`;
        const StyledE = styled("div") \`\${() => someFunc()}\`;
        const StyledF = styled("div") \`\${someVal}\`;


        <StyledA />;
        <StyledB />;
        <StyledC />;
        <StyledD />;
        <StyledE />;
        <StyledF />;
        `;
    expect(getMocks(source, "flow")).toMatchSnapshot();
});

it("Does not import inside in tagged template 2", () => {
    const source = `
        import { A, B } from "some-lib";
        import styled, { theme } from "styled";

        const StyledA = styled(A)\`
            color: \${theme.color.black};
        \`;

        const StyledB = styled(B)\`
            color: \${theme.color.white};
        \`;

        <StyledA />;
        <StyledB />;
        `;
    expect(getMocks(source, "flow")).toMatchSnapshot();
});

it("Works when HOC component was used in HOC for other component", () => {
    const source = `
        import A from "a";
        import styled from "styled";

        const StyledA = styled(A)\`
            font-size: 1em;
        \`;
        const StyledB = styled(StyledA)\`
            width: 50px;
        \`;
        const StyledC = styled(StyledB)\`
            height: 60px;
        \`;

        const StyledD = HOC(StyledA);
        const StyledE = HOC({}, StyledB);
        const StyledF = HOC(StyledE, {});

        <StyledA />;
        <StyledB />;
        <StyledC />;
        <StyledD />;
        <StyledE />;
        <StyledF />;
        `;
    expect(getMocks(source, "flow")).toMatchSnapshot();
});

it("Works for nested member expressions inside HOC", () => {
    const source = `
        import A from "a";
        import styled from "styled";

        const Styled = styled(A.B.C);

        <Styled />;
        `;
    expect(getMocks(source, "flow")).toMatchSnapshot();
});

it("Works for member expression without HOC", () => {
    const source = `
        import A from "a";
        import * as B from "b";

        const FakeA = A.B;
        const FakeB = B.C.D;

        <FakeA />;
        <FakeB />;
        `;
    expect(getMocks(source, "flow")).toMatchSnapshot();
})

it("Works with TS annotations", () => {
    const source = `
        import { A } from "a";
        import styled from "styled";

        const StyledA: SomeComponentClass = A;
        <StyledA />;

        const StyledB: SomeClass<Props> = styled(A);
        <StyledB />;

        const StyledC = styled<Props>(A)\`\`;
        <StyledC />;

        const StyledD = (styled as any)(A);
        <StyledD />;

        const StyledE = styled(A as SomeClass)\`\`;
        <StyledE />;

        const StyledF = styled(A) as any as SomeClass<Props>;
        <StyledF />;

        const StyledG = styled.withComponent(A);
        <StyledG />;

        const StyledH = styled.withComponent<Props>(A)\`\`;
        <StyledH />;

        const StyledK = styled.withComponent(A as SomeClass<Props>)\`\`;
        <StyledK />;
        `;
    expect(getMocks(source, "typescript")).toMatchSnapshot();
});

it("Real source 1", () => {
    const source = `
        import { ModalCloseIcon, ModalHeader } from "@ramble/ramble-ui";
        import React from "react";
        import styled, { theme } from "../../theme";

        const MapHeader = styled(ModalHeader) \`
            padding-bottom: 0.5rem;
            border-bottom: 0;

            /* since modal will be basic and dimmed we need contrast color of dimmmer */
            color: \${theme.getContrastTextColor(theme.dimmer.black).string()};
            justify-content: space-between;
        \`;

        const CloseIcon = styled(ModalCloseIcon) \`
            color: \${theme.getContrastTextColor(theme.dimmer.black).string()};
            padding: 0;
            margin: 0;
        \`;

        <MapHeader />;
        <CloseIcon />;
        `;
    expect(getMocks(source, "flow")).toMatchSnapshot();
})

it("Real source 2", () => {

    const source = `
        import Scrollbar, { ScrollbarProps } from "react-smooth-scrollbar";
        import { AreaWrapperContainer, Message, MessagesContentContainer, PresenceMessage } from "../Elements";

        const StyledScrollbar = MessagesContentContainer.withComponent(Scrollbar) as any as StyledComponent<ScrollbarProps & { innerRef: (r: Scrollbar | null) => void }>;

        <StyledScrollbar />;
        `;
    expect(getMocks(source, "typescript")).toMatchSnapshot();
});

it("Real source 3", () => {
    const source = `
        "use strict";

        import React from "react";
        import styled from "../../../theme";
        import { HeaderDropdownMenu } from "../HeaderDropdownMenu";

        // Like processed by emotion/styled-components babel plugin
        const MenuItem =
        /*#__PURE__*/
        styled(HeaderDropdownMenu, {
        label: "MenuItem",
        target: "ev6xfg12"
        })("padding:11px 16px;width:100%;display:flex;flex-flow:column nowrap;justify-content:flex-start;align-items:stretch;& > .toggle{margin-left:1em;}");

        function TeamStatusItem(props) {
            return (
                <MenuItem />
            );
        }
        `;
    expect(getMocks(source, "flow")).toMatchSnapshot();
});

it("Works for alwaysMockIdentifiers without JSX opening element", () => {
    const source = `
        import A from "a";
        import { B } from "../b";
        import C from "c";
        import D from "d";
        import styled from "styled";

        const StyledA = styled(A)\`\`;
        const StyledB = styled(A.B);
        const StyledC = styled(A.C, {});
        const StyledD = styled(A.D, {})\`font-size: 8em;\`;

        const B1 = A.withComponent(C.C1);
        const B2 = styled(D) \`\`;

        const B3 = nonStyled(C.C2);
        `;
    expect(getMocks(source, "flow")).toMatchSnapshot();
});
