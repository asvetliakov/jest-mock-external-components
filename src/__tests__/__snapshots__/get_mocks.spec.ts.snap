// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Does not append imported if not used in HOC as argument 1`] = `
Array [
  Object {
    "identifier": "A",
    "path": "a",
    "type": "default",
  },
  Object {
    "identifier": "C.A.B",
    "path": "../c",
    "type": "default",
  },
]
`;

exports[`Does not import inside in tagged template 1`] = `
Array [
  Object {
    "identifier": "A",
    "path": "a",
    "type": "default",
  },
]
`;

exports[`Does not import inside in tagged template 2 1`] = `
Array [
  Object {
    "identifier": "A",
    "path": "some-lib",
    "type": "name",
  },
  Object {
    "identifier": "B",
    "path": "some-lib",
    "type": "name",
  },
]
`;

exports[`Real source 1 1`] = `
Array [
  Object {
    "identifier": "ModalHeader",
    "path": "@ramble/ramble-ui",
    "type": "name",
  },
  Object {
    "identifier": "ModalCloseIcon",
    "path": "@ramble/ramble-ui",
    "type": "name",
  },
]
`;

exports[`Real source 2 1`] = `
Array [
  Object {
    "identifier": "Scrollbar",
    "path": "react-smooth-scrollbar",
    "type": "default",
  },
]
`;

exports[`Real source 3 1`] = `
Array [
  Object {
    "identifier": "HeaderDropdownMenu",
    "path": "../HeaderDropdownMenu",
    "type": "name",
  },
]
`;

exports[`Works for HOC like components 1`] = `
Array [
  Object {
    "identifier": "A",
    "path": "a",
    "type": "default",
  },
  Object {
    "identifier": "A.B",
    "path": "a",
    "type": "default",
  },
  Object {
    "identifier": "A.C",
    "path": "a",
    "type": "default",
  },
  Object {
    "identifier": "A.D",
    "path": "a",
    "type": "default",
  },
]
`;

exports[`Works for JSX expressions 1`] = `
Array [
  Object {
    "identifier": "A",
    "path": "a",
    "type": "default",
  },
  Object {
    "identifier": "A.A1",
    "path": "a",
    "type": "default",
  },
  Object {
    "identifier": "B",
    "path": "../b",
    "type": "name",
  },
  Object {
    "identifier": "B.B1",
    "path": "../b",
    "type": "name",
  },
  Object {
    "identifier": "C.C1",
    "path": "c/test",
    "type": "namespace",
  },
  Object {
    "identifier": "C.C1.C2",
    "path": "c/test",
    "type": "namespace",
  },
  Object {
    "identifier": "C.C2.C3.C4",
    "path": "c/test",
    "type": "namespace",
  },
]
`;

exports[`Works for alwaysMockIdentifiers without JSX opening element 1`] = `
Array [
  Object {
    "identifier": "A",
    "path": "a",
    "type": "default",
  },
  Object {
    "identifier": "A.B",
    "path": "a",
    "type": "default",
  },
  Object {
    "identifier": "A.C",
    "path": "a",
    "type": "default",
  },
  Object {
    "identifier": "A.D",
    "path": "a",
    "type": "default",
  },
  Object {
    "identifier": "C.C1",
    "path": "c",
    "type": "default",
  },
  Object {
    "identifier": "D",
    "path": "d",
    "type": "default",
  },
]
`;

exports[`Works for member expression without HOC 1`] = `
Array [
  Object {
    "identifier": "A.B",
    "path": "a",
    "type": "default",
  },
  Object {
    "identifier": "B.C.D",
    "path": "b",
    "type": "namespace",
  },
]
`;

exports[`Works for nested member expressions inside HOC 1`] = `
Array [
  Object {
    "identifier": "A.B.C",
    "path": "a",
    "type": "default",
  },
]
`;

exports[`Works when HOC component was used in HOC for other component 1`] = `
Array [
  Object {
    "identifier": "A",
    "path": "a",
    "type": "default",
  },
]
`;

exports[`Works with TS annotations 1`] = `
Array [
  Object {
    "identifier": "A",
    "path": "a",
    "type": "name",
  },
]
`;
